Main menu and tool code refactoring

kvakvs opened the pull request:
# What does

Affects Main TMPE button, and main TMPE panel

![bild](https://user-images.githubusercontent.com/288724/74793879-01780f80-52c2-11ea-8ca0-2cfcb7d51080.png)

* Main panel and main button now scale to % of screen size
* button states all reworked, normal sprited desaturated, active sprites full colour
* textures split into sprites creating atlas on load
* smart skinning per button, and buttons also control their state

# More details

* `BaseUI` and `UIBase` renamed to `ModUI` (class) and `ModUi` (field) representing root object of user interface for the TMPE mod;
* Some generic UI code is now slowly moving out to `TrafficManager.U` namespace (`BaseUButton` from old `LinearSpriteButton`)
* WIP on `BaseUButton` and `ButtonTexture` going to generalize textures handling from `UIMainMenuButton`
* Get-only properties on `BaseUButton` renamed to `Get*` functions instead of just nouns.
* New texture loading code for atlases is used for UI buttons, removing some loading code from the old static texture classes.
* Renaming and documenting stuff.
* Logic is mostly unchanged, but file structure can be changing, files are moved or split.

This is code cleanup and preparation for new main menu panel design.

kianzarrin commented on TLM/TLM/U/BaseUButton.cs:
```diff
@@ -0,0 +1,109 @@
+            playAudioEvents = true;
+        }
+
+        public abstract bool Active { get; }
```
Note: My road selection also provides an optional disabled state:
https://github.com/kianzarrin/Cities-Skylines-Traffic-Manager-President-Edition/blob/cb19aa340e2f6e22e24f1bca015a0191ea84ba80/TLM/TLM/UI/LinearSpriteButton.cs#L27

kvakvs replied:
Yeah this entire thing is convoluted, i want a more generic button with all these states, now also including the disabled state. Will have remember to take your change in too, whether your or mine gets merged first (your possibly goes in first).

aubergine10 commented:
Another state for buttons that might be useful is some sort of "highlight" effect?

Currently there's no visual feedback to users as to what bulk applicator tools are applying; usually it is the currently selected tool + one or more other tools.

For example, with priority signs tool selected, I can Ctrl+Shift+Click to apply priority signs, junction restrictions and lane arrows down a route. Would be nice if the toolbar buttons of those tools could light a different colour (eg. light blue bg to match the selection shape on map) letting me know I'm about to apply multiple tools.

At some point in future, it would also be nice to toggle persistent overlays by right-clicking buttons so maybe also an "overlay" state?

kvakvs commented:
I like the ideas but not going to do any of that until a clear state diagram is created.
Even if a possibility of state toggles like those right click overlays, we can add that later to the design, let's begin with just refactoring + changing textures for the main menu.

kvakvs commented on TLM/TLM/U/Button/ButtonSkin.cs:
```diff
@@ -0,0 +1,150 @@
+            }
+            if (BackgroundActive) {
+                names.Add($"{BackgroundPrefix}-bg-active");
+            }
```
Also check BackgroundPrefix is null or empty

kvakvs commented on TLM/TLM/UI/MainMenu/MainMenuPanel.cs:
```diff
@@ -160,61 +142,81 @@ public class SizeProfile {
+            int y = (int)GetScaledTitlebarHeight();
+            float buttonSize = GetScaledButtonSize();
+            const int numRows = 2;
+            const int numCols = 6;
```
Rework, move constants

kvakvs commented on TLM/TLM/UI/ModUI.cs:
```diff
@@ -2,12 +2,43 @@ namespace TrafficManager.UI {
+        /// <summary>
+        /// List of main menu button functions, used for loading textures.
+        /// </summary>
+        public List<string> ButtonList = new List<string> {
```
Double check if this is used

kvakvs replied:
Not used

aubergine10 commented:
> ModUi (field)

As in a reference to main instance of `ModUI` class? if so, would `ModUI.Instance` be better?

```cs
public static ModUI instance;

public static ModUI Instance => instance ?? (instance = new ModUI());
```

?

kvakvs commented:
Possibly you are right.
But that would decentralize copies of classes in static globals, pushing us away from the golden future of no singletons.
It would make the code look better, i can agree.
And i still need to control the time of its creation, i.e. it can't be accessible before its created, and it can't be created before we're in the game.

aubergine10 commented:
The submodules might need updating to reflect changes of #664

aubergine10 commented:
TMPE button and the main menu seem larger than usual for me:

![image](https://user-images.githubusercontent.com/1386719/74807867-6c850e80-52e1-11ea-8470-54edd39d49f0.png)

![image](https://user-images.githubusercontent.com/1386719/74807927-8de5fa80-52e1-11ea-8eb3-c0bea899d349.png)

My screen resolution: 2560 x 1440

Suggestions:

1. The TMPE button should be same size as Info Views button
    * Note that mods such as Hide It can remove common buttons so they might not be available for reference
2. Could we get a slider in mod options to tweak the scaling factor? I like smaller than usual UI so I can see as much of the map as possible, others might light larger than usual UI.

kvakvs commented:
Main button size is defined as % of screen size based on its old pixel size and reference resolution 1920x1080, it is scaled to current screen width or height whichever is smallest.
Smaller main button: I can tune down the %'s used so it becomes 40px instead of 50px in 1080p.
Slider: Yes we can have a slider
Will the slider affect all UI windows AND the main button? In that case the problem is solved by having that slider.

aubergine10 commented:
> Smaller main button: I can tune down the %'s used so it becomes 40px instead of 50px in 1080p

> Is it better to have fixed size for the button and then use that slider?

Default scaling should be set in such a way as to make TMPE button same size as Info Views button, whatever that may be for current screen resolution.

> Will the slider affect all UI windows AND the main button? In that case the problem is solved by having that slider.

Not sure. Personally I sort of position the TMPE button up near top alongside the similarly-designed vanilla buttons. So it would look a bit weird if scaling also affected the TMPE button. It might be a case of trial and error to see what works best. But personally I think the scaling should affect all our UI with the exception of the TMPE button.

_Also, at some point, it would be nice to make the TMPE button adopt the styling/sizing of where the user places it on screen. For example, if it's over the vanilla toolbar strip (roads, zoning, districts....) it could take on the styling/sizing of those buttons. If it's down at the very bottom (time bar, RCI chart, etc) it could take on styling to fit better in that area._

kvakvs commented:
OK, next plan then is:
Figure out on-screen button sizing
UI scale slider (will only affect main panel now but in the perspective all mod's UI panels)
i would not bother with further styling in this ticket, let FireController1847 do his research, but we have the tech to load multiple and switch skins on the fly.

aubergine10 commented:
> we have the tech to load multiple and switch skins on the fly.

![image](https://user-images.githubusercontent.com/1386719/74825805-723f1c00-5302-11ea-92e1-42ddfe864b05.png)

kvakvs commented:
The scaling is not an issue, it just scales to the same % of screen size that looks good on my screen :D Problem is that other buttons don't scale so it looks out of place. I'll see how to fix that now, remove scaling probably.

aubergine10 commented:
![image](https://user-images.githubusercontent.com/1386719/74883220-61c18c80-5368-11ea-8364-3258ef54f5c0.png)

kvakvs commented:
Fixed main button to always 50px.
Added min scale for main panel button to be no smaller than 40px, but possibly larger.

krzychu124 commented on TLM/TLM/UI/MainMenu/MainMenuPanel.cs:
```diff
@@ -160,61 +142,90 @@ public class SizeProfile {
+            internal static float GetHeight() {
+                // 2 button rows + spacing (1/8th) + titlebar
+                return (GetButtonSize() * 2.125f) + GetTitlebarHeight();
+            }
```
I think you forgot to add spacings(right and bottom) because it looks like this (ModTools debugView):

![image](https://user-images.githubusercontent.com/19638970/74887863-79a30b80-537c-11ea-8591-a24c5ac8c387.png)

kvakvs replied:
Added missing spacing right and bottom

kvakvs commented:
Remaining: UI scale slider.

krzychu124 commented:
So far looks very good :)

aubergine10 commented:
Assuming all the TM:PE panels will use some new panel class (or whatever) can we ensure the panels soak up mouse clicks/drags to prevent things like #683

kvakvs commented:
@aubergine10 Happy to do that, but i am forgetful. please remind me :)

kvakvs commented:
Added UI slider for UI scale, which applies in real time as you drag it.

aubergine10 commented:
> was it IMGUI based window which leaked mouse clicks? It should not happen if i refactor IMGUI to normal CO.UI

It was the Speed Limits panel that was letting clicks/drags pass through to the TM:PE main menu (which was maybe using the older UI system?)

kvakvs commented:
Despawn button is confusing. The logic looks fine to me:
if disable despawning: tooltip becomes "Enable despawning" and it highlights blue
if not disable despawning: tooltip becomes "Disable despawning" and highlight is removed
But the behaviour in game looked inverted?

aubergine10 commented:
I found the tooltip confusing. When there's a toggle state, IMO the tootip should just clarify what the _current_ state is, for example:

> Despawning is currently: Enabled  
> Disabling despawning makes traffic more realistic, but harder to manage.

kvakvs commented:
Strings added to Crowdin but not updated/tested yet.

aubergine10 commented:
I have a bunch of feedback on keybinds and tooltips, but I think that is best left for separte PR; I will create an issue with that feedback.

aubergine10 commented:
While glancing through code I noticed there was some stuff in there about arranging stuff in to rows/columns.... Could we imrpove how the menu reacts to tools being enabled/disabled in "Maintenance" tab of mod settings?

Because if I turn all the the tools off, this is what I get:

![image](https://user-images.githubusercontent.com/1386719/74894608-68abc780-5387-11ea-8637-6406db946e44.png)

kianzarrin commented:
![](https://user-images.githubusercontent.com/288724/74793879-01780f80-52c2-11ea-8ca0-2cfcb7d51080.png)
The number 11 initially made me think there are 11 new things to look at. like this:
![](https://linustechtips.com/main/uploads/monthly_2019_03/Notifications.jpg.bfa6203d004dc0d8a55b72d5f93afb77.jpg)
maybe put V11 if it fits?

aubergine10 commented:
The `11` is already in title bar of the TMPE menu bar. So I think it can be removed from the menu button.

kvakvs commented:
11 is small and easy to remove. Reason it is here, because i wanted the button to look slightly different from pre-11, all to spark interests in those who still run v10 and didn't pay attention that TMPE has been upgraded to 11. One idea would be to take whatever we have for logo? The chirper and traffic light? And take away the crown. And another idea was to add version to it.

kvakvs commented on TLM/TLM/UI/MainMenu/MainMenuPanel.cs:
```diff
@@ -255,5 +267,9 @@ public class SizeProfile {
+
+        // TODO: Scale with screen size in a smart way
+        public static float GetButtonWidth() => 50f;
+        public static float GetButtonHeight() => 50f;
```
Remove this

kvakvs commented on TLM/TLM/UI/MainMenu/VehicleRestrictionsButton.cs:
```diff
@@ -1,13 +1,27 @@
+                                                      BackgroundActive = true,
+                                                      ForegroundActive = true,
+                                                  };
+            atlasKeys.AddRange(this.Skin.CreateAtlasKeysList());
```
Ensure that keys added to the atlas are unique. Got a feeling that background is added every time.

kvakvs commented on TLM/TLM/UI/RemoveCitizenInstanceButtonExtender.cs:
```diff
@@ -49,11 +52,11 @@ public class RemoveCitizenInstanceButtonExtender : MonoBehaviour {
         }
 
-        public class RemoveCitizenInstanceButton : LinearSpriteButton {
+        public class RemoveCitizenInstanceButton : BaseUButton {
```
Double check this still works and the icon is visible.

kvakvs commented on TLM/TLM/UI/RemoveVehicleButtonExtender.cs:
```diff
@@ -57,11 +60,11 @@ var publicTransportVehicleInfoPanel
         }
 
-        public class RemoveVehicleButton : LinearSpriteButton {
+        public class RemoveVehicleButton : BaseUButton {
```
Double check this works too, and the button is visible

krzychu124 commented:
Scale slider works as intended, saves position. looks good 👍

kvakvs commented:
Now with maintenance features disabled, the menu is refreshed live and buttons are removed. As soon as a row is empty, the panel also shrinks.
![bild](https://user-images.githubusercontent.com/288724/74998934-6fefd580-545a-11ea-829d-10c0c60aa4ab.png)

kvakvs commented:
Updated now:
* Code (toolbutton atlases now use HashSet<string> instead of List<string> to fight duplicate sprite names)
* Buttons updated in realtime as Maintenance tab has its features toggled
* Added v11 to the main button sprite, and toned down the crown in the background.

aubergine10 commented:
In `DEBUG` builds, `ModUI.cs`:

![image](https://user-images.githubusercontent.com/1386719/75029318-be709480-5499-11ea-838b-be68b21dca2e.png)

> The name `uiView` does not exist in the current context

kvakvs commented:
@aubergine10 
Fixed, i've commented too much last time

aubergine10 commented:
Tested out menu resizing when tools are disabled...

![image](https://user-images.githubusercontent.com/1386719/75030044-2ecbe580-549b-11ea-8354-8b9d7689cde3.png)

![image](https://user-images.githubusercontent.com/1386719/75030066-3e4b2e80-549b-11ea-84f8-701bc58fbd44.png)

![image](https://user-images.githubusercontent.com/1386719/75030173-781c3500-549b-11ea-88eb-da52c3c591a6.png)

Is it possible to tweak so it works something like:

```js
// pseudocode
if (numButtons > 4) {
   // arrange buttons on to two rows
   // but retain at least 4 buttons on top row
} else {
   // just one row with 4 or less buttons
}
```

kvakvs commented:
Possible, but reducing width will affect the text title label. At 6 buttons the window becomes 3 buttons wide and label will stick beyond the window. So your number should be at least 6, not 4. And with Kian's debug "HOT RELOAD" extra long label it will hang off the window for a mile.

aubergine10 commented:
I think with the hot reload it doesn't matter too much as only devs see that.

Maybe min 5 icon width for top row would be better (assuming that fits STABLE version label)?

kvakvs commented:
I'll go with your original idea, didn't notice you had more complex logic there than just 4 button count, it was 4 buttons on top row which should be enough to fit version label.

kvakvs commented:
I've made a layout class which handles button placement and row breaks, it is not tested, i am at work and have no way to run the game. If you wish to play the happy debugging, you can take it now, otherwise i welcome you to try this branch later tonight.

aubergine10 commented:
Tested in-game - the button rows/cols is working great. However the panel backgorund is glitching.

![image](https://user-images.githubusercontent.com/1386719/75038566-858eea80-54ae-11ea-9e26-dffb97bb2bf9.png)

Panel background didn't update when adding 5th button:

![image](https://user-images.githubusercontent.com/1386719/75038618-9f303200-54ae-11ea-9aa4-168fab8d44bd.png)

At 8 buttons added it looks like panel background is being extended wrong way (or think there's extra row?)

![image](https://user-images.githubusercontent.com/1386719/75038696-cc7ce000-54ae-11ea-8cba-c752bdb6d198.png)

9 buttons and panel height is correct but width still only 4 buttons wide:

![image](https://user-images.githubusercontent.com/1386719/75038731-e8808180-54ae-11ea-9a63-aa2cd1646e4a.png)

10 buttons, panel bg too high again:

![image](https://user-images.githubusercontent.com/1386719/75038754-f7ffca80-54ae-11ea-904f-55d5de677a45.png)

etc...
